src/Typecast/SimpleTypecast.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Hydrator\Typecast;
+
+use ReflectionClass;
+use ReflectionNamedType;
+use ReflectionType;
+use ReflectionUnionType;
+use Stringable;
+use Yiisoft\Hydrator\Hydrator;
+use Yiisoft\Hydrator\SkipTypecastException;
+use Yiisoft\Hydrator\TypecastInterface;
+use Yiisoft\Strings\NumericHelper;
+
+use function is_array;
+use function is_bool;
+use function is_float;
+use function is_int;
+use function is_null;
+use function is_object;
+use function is_string;
+
+final class SimpleTypecast implements TypecastInterface
+{
+    public function cast(mixed $value, ?ReflectionType $type, Hydrator $hydrator): mixed
+    {
+        if ($type instanceof ReflectionNamedType) {
+            $types = [$type];
+        } elseif ($type instanceof ReflectionUnionType) {
+            $types = array_filter(
+                $type->getTypes(),
+                static fn(mixed $type) => $type instanceof ReflectionNamedType,
+            );
+        } elseif ($type === null) {
+            return $value;
+        } else {
+            throw new SkipTypecastException();
+        }
+
+        foreach ($types as $t) {
+            if ($t->isBuiltin()) {
+                if ($t->allowsNull() && is_null($value)) {
+                    return null;
+                }
+                switch ($t->getName()) {
```

Comment:
```suggestion
                match ($t->getName()) {
```

Reply:
Seems, it doesn't fit here. How you see usage `match` here?

Reply:
Oh yeah, `match` doesn't support multiline expressions.